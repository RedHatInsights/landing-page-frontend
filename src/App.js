import React, {
  useEffect,
  useState,
  createContext,
  lazy,
  Suspense,
} from 'react';
import { withRouter, Switch, Route } from 'react-router-dom';
import { getRegistry } from '@redhat-cloud-services/frontend-components-utilities/files/cjs/Registry';
import { useDispatch } from 'react-redux';
import { Bullseye, Spinner } from '@patternfly/react-core';
import technologiesReducer from './store/technologiesReducer';
import contentStore from './store/contentReducer';
import {
  technologiesLoaded,
  loadEndpoints,
  loadCarousel,
  loadSections,
} from './store/actions';
import { activeTechnologies } from './consts';

const Landing = lazy(() =>
  import(/* webpackCunkName: "Landing" */ './routes/Landing')
);
const Maintenance = lazy(() =>
  import(/* webpackCunkName: "Maintenance" */ './routes/Maintenance')
);
const NotFound = lazy(() =>
  import(/* webpackCunkName: "NotFound" */ './routes/404')
);

import './App.scss';

const routes = {
  landing: '/',
  maintenance: '/maintenance',
};

export const PermissionContext = createContext();

const App = () => {
  const [isOrgAdmin, setIsOrgAdmin] = useState();
  const dispatch = useDispatch();

  useEffect(() => {
    getRegistry().register({ technologies: technologiesReducer, contentStore });
    dispatch(technologiesLoaded(activeTechnologies));
    const endpointsAction = loadEndpoints();
    dispatch(endpointsAction);
    (async () => {
      await endpointsAction.payload;
      dispatch(loadCarousel());
      dispatch(loadSections());
    })();
    insights.chrome.init();
    insights.chrome.identifyApp('landing');
    window.insights.chrome.auth
      .getUser()
      .then((user) => user && setIsOrgAdmin(user.identity.user.is_org_admin));
  }, []);

  return (
    <PermissionContext.Provider value={{ isOrgAdmin }}>
      <Suspense
        fallback={
          <Bullseye>
            <Spinner size="xl" />
          </Bullseye>
        }
      >
        <Switch>
          <Route exact path={routes.landing} component={Landing} />
          <Route exact path={routes.landingBeta} component={Landing} />
          <Route exact path={routes.maintenance} component={Maintenance} />
          <Route path="*" component={NotFound} />
        </Switch>
      </Suspense>
    </PermissionContext.Provider>
  );
};

export default withRouter(App);
