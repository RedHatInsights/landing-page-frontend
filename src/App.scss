@import '~@patternfly/patternfly/sass-utilities/_all';

$pf-c-width-limiter--breakpoint-map: build-breakpoint-map();

// width limiter
.pf-c-width-limiter {
    --pf-c-width-limiter--MaxWidth: auto;
    --pf-c-width-limiter--MinWidth: 0;

    @include pf-build-css-variable-stack(
        "max-width",
        "--pf-c-width-limiter--MaxWidth",
        $pf-c-width-limiter--breakpoint-map
    );
    @include pf-build-css-variable-stack(
        "min-width",
        "--pf-c-width-limiter--MinWidth",
        $pf-c-width-limiter--breakpoint-map
    );
}

.pf-c-page__main-section {
    --ins-c-page__main-section--PaddingRight: var(--pf-global--spacer--xl);
    --ins-c-page__main-section--PaddingLeft: var(--pf-global--spacer--xl);
    --ins-c-landing--m-display-md--MaxWidth: #{pf-size-prem(1283px)};
    --ins-c-landing--m-display-lg--MaxWidth: #{pf-size-prem(1600px)};

    @media screen and (min-width: $pf-global--breakpoint--md) {
        --ins-c-page__main-section--PaddingRight: var(--pf-global--spacer--2xl);
        --ins-c-page__main-section--PaddingLeft: var(--pf-global--spacer--2xl);
    }

    --pf-c-page__main-section--PaddingRight: var(--ins-c-page__main-section--PaddingRight);
    --pf-c-page__main-section--PaddingLeft: var(--ins-c-page__main-section--PaddingLeft);

    &.pf-m-limit-width.ins-m-display-md {
        --pf-c-page--section--m-limit-width--MaxWidth: calc(var(--ins-c-landing--m-display-md--MaxWidth) + var(--pf-c-page__main-section--PaddingRight) + var(--pf-c-page__main-section--PaddingLeft));
    }

    &.pf-m-limit-width.ins-m-display-lg {
        --pf-c-page--section--m-limit-width--MaxWidth: var(--ins-c-landing--m-display-lg--MaxWidth);
    }

    &.pf-m-center .pf-c-page__main-body {
        width: 100%;
        margin-right: auto;
        margin-left: auto;
    }
}

.pf-c-page__main-section.ins-m-primary-color--200 {
    --pf-c-page__main-section--BackgroundColor: var(--pf-global--primary-color--200);

    color: var(--pf-global--Color--light-100);
}

// temp styles
.pf-c-drawer__content,
.ins-m-full-height {
    display: flex;
    flex-direction: column;
    height: 100%;
}

.pf-c-content {
    --pf-c-content--ul--PaddingLeft: var(--pf-global--spacer--xs);
    --pf-c-content--ul--MarginLeft: var(--pf-global--spacer--md);
}

.ins-c-logo {
    --ins-c-logo--Width: #{pf-size-prem(160px)};

    width: var(--ins-c-logo--Width);
}

.pf-l-grid {
    > * {
        max-width: 100%;
    }

    &.pf-m-align-items-center {
        align-items: center;
    }

    .pf-l-grid__item {
        &.pf-m-justify-items-center {
            justify-items: center;
        }
    }
}

.ins-c-card-video {
    position: relative;
    display: grid;
    row-gap: var(--pf-global--spacer--lg);
    width: 100%;

    @media screen and (min-width: 537px) {
        width: 473px;
    }
}

.ins-c-card-video__iframe {
    position: relative;
    padding-top: 56.25%;

    > iframe {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
    }
}

.ins-c-card-video__iframe,
.ins-c-card-video__figcaption {
    width: 100%;
}
