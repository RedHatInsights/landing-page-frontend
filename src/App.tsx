import React, {
  Suspense,
  createContext,
  lazy,
  useEffect,
  useState,
} from 'react';
import '@patternfly/patternfly/utilities/Text/text.css';
import { Route, Routes } from 'react-router-dom';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import { Bullseye } from '@patternfly/react-core/dist/dynamic/layouts/Bullseye';
import { Spinner } from '@patternfly/react-core/dist/dynamic/components/Spinner';
import { Provider } from 'react-redux';
import { getRegistry } from '@redhat-cloud-services/frontend-components-utilities/Registry';
import promiseMiddleware from 'redux-promise-middleware';
import notificationsMiddleware from '@redhat-cloud-services/frontend-components-notifications/notificationsMiddleware';
import ReducerRegistry from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { Middleware, Reducer } from 'redux';
import NotificationsPortal from '@redhat-cloud-services/frontend-components-notifications/NotificationPortal';
import { notificationsReducer } from '@redhat-cloud-services/frontend-components-notifications/redux';
import { useNavigate } from 'react-router-dom';
import logger from 'redux-logger';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let registry: ReducerRegistry<any>;

export function init(...middleware: Middleware[]) {
  registry = getRegistry({}, [
    promiseMiddleware,
    notificationsMiddleware({ errorDescriptionKey: ['detail', 'stack'] }),
    ...middleware,
  ]);
  return registry;
}

const Landing = lazy(
  () => import(/* webpackCunkName: "Landing" */ './routes/Landing')
);
const Maintenance = lazy(
  () => import(/* webpackCunkName: "Maintenance" */ './routes/Maintenance')
);
const NotFound = lazy(
  () => import(/* webpackCunkName: "NotFound" */ './routes/404')
);

const routes = {
  landing: '/',
  maintenance: '/maintenance',
};

export const PermissionContext = createContext<{
  isOrgAdmin?: boolean;
}>({});

const App = () => {
  const [isOrgAdmin, setIsOrgAdmin] = useState<boolean>(false);
  const chrome = useChrome();
  const navigate = useNavigate();
  const { on } = useChrome();

  chrome?.updateDocumentTitle?.('Hybrid Cloud Console Home', false);

  useEffect(() => {
    chrome.auth
      .getUser()
      .then(
        (user) => user && setIsOrgAdmin(!!user?.identity?.user?.is_org_admin)
      );
  }, []);

  useEffect(() => {
    const registry = getRegistry();
    registry.register({ notifications: notificationsReducer as Reducer });

    const unregister = on('APP_NAVIGATION', (event) =>
      navigate(`/${event.navId}`)
    );
    return () => {
      unregister?.();
    };
  }, []);

  return (
    <Provider
      store={init(
        ...(process.env.NODE_ENV !== 'production' ? [logger] : [])
      ).getStore()}
    >
      <PermissionContext.Provider value={{ isOrgAdmin }}>
        <NotificationsPortal />
        <Suspense
          fallback={
            <Bullseye>
              <Spinner size="xl" />
            </Bullseye>
          }
        >
          <Routes>
            <Route path={routes.landing} element={<Landing />} />
            <Route path={routes.maintenance} element={<Maintenance />} />
            <Route path="*" element={<NotFound />} />
          </Routes>
        </Suspense>
      </PermissionContext.Provider>
    </Provider>
  );
};

export default App;
