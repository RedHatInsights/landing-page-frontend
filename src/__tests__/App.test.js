import React from 'react';
import { act } from 'react-dom/test-utils';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import App from '../App';

const mockStore = configureMockStore();
const store = mockStore({});

jest.mock('../routes/Landing', () => ({
  __esModule: true,
  // eslint-disable-next-line react/display-name
  default: () => <span>Landing</span>,
}));

describe('App component', () => {
  it('should render correctly', async () => {
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <Provider store={store}>
          <Router>
            <App />
          </Router>
        </Provider>
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
