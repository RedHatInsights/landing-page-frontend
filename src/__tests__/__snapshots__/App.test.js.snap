// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <withRouter(App)>
        <App
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <Suspense
            fallback={
              <Bullseye>
                <Spinner
                  size="xl"
                />
              </Bullseye>
            }
          >
            <Switch>
              <Route
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": Promise {},
                    "_status": 0,
                  }
                }
                computedMatch={
                  Object {
                    "isExact": true,
                    "params": Object {},
                    "path": "/",
                    "url": "/",
                  }
                }
                exact={true}
                location={
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  }
                }
                path="/"
              />
            </Switch>
            <Bullseye>
              <div
                className="pf-l-bullseye"
              >
                <Spinner
                  size="xl"
                >
                  <span
                    aria-valuetext="Loading..."
                    className="pf-c-spinner pf-m-xl"
                    role="progressbar"
                  >
                    <span
                      className="pf-c-spinner__clipper"
                    />
                    <span
                      className="pf-c-spinner__lead-ball"
                    />
                    <span
                      className="pf-c-spinner__tail-ball"
                    />
                  </span>
                </Spinner>
              </div>
            </Bullseye>
          </Suspense>
        </App>
      </withRouter(App)>
    </Router>
  </BrowserRouter>
</Provider>
`;

exports[`App component should render correctly with data 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <withRouter(App)
        technologies={
          Object {
            "foo": "bar",
          }
        }
      >
        <App
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          technologies={
            Object {
              "foo": "bar",
            }
          }
        >
          <Suspense
            fallback={
              <Bullseye>
                <Spinner
                  size="xl"
                />
              </Bullseye>
            }
          >
            <Switch>
              <Route
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": [Function],
                    "_status": 1,
                  }
                }
                computedMatch={
                  Object {
                    "isExact": true,
                    "params": Object {},
                    "path": "/",
                    "url": "/",
                  }
                }
                exact={true}
                location={
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  }
                }
                path="/"
              >
                <withRouter(Landing)
                  history={
                    Object {
                      "action": "POP",
                      "block": [Function],
                      "createHref": [Function],
                      "go": [Function],
                      "goBack": [Function],
                      "goForward": [Function],
                      "length": 1,
                      "listen": [Function],
                      "location": Object {
                        "hash": "",
                        "pathname": "/",
                        "search": "",
                        "state": undefined,
                      },
                      "push": [Function],
                      "replace": [Function],
                    }
                  }
                  location={
                    Object {
                      "hash": "",
                      "pathname": "/",
                      "search": "",
                      "state": undefined,
                    }
                  }
                  match={
                    Object {
                      "isExact": true,
                      "params": Object {},
                      "path": "/",
                      "url": "/",
                    }
                  }
                >
                  <Landing
                    history={
                      Object {
                        "action": "POP",
                        "block": [Function],
                        "createHref": [Function],
                        "go": [Function],
                        "goBack": [Function],
                        "goForward": [Function],
                        "length": 1,
                        "listen": [Function],
                        "location": Object {
                          "hash": "",
                          "pathname": "/",
                          "search": "",
                          "state": undefined,
                        },
                        "push": [Function],
                        "replace": [Function],
                      }
                    }
                    loadTechnologies={[Function]}
                    location={
                      Object {
                        "hash": "",
                        "pathname": "/",
                        "search": "",
                        "state": undefined,
                      }
                    }
                    match={
                      Object {
                        "isExact": true,
                        "params": Object {},
                        "path": "/",
                        "url": "/",
                      }
                    }
                  >
                    <Loading>
                      <Connect(Main)>
                        <Main>
                          <section
                            className="pf-l-page__main-section pf-c-page__main-section"
                            page-type=""
                          >
                            <Spinner
                              centered={true}
                            >
                              <div
                                className="ins-c-spinner ins-m-center"
                                role="status"
                              >
                                <span
                                  className="pf-u-screen-reader"
                                >
                                  Loading...
                                </span>
                              </div>
                            </Spinner>
                          </section>
                        </Main>
                      </Connect(Main)>
                    </Loading>
                  </Landing>
                </withRouter(Landing)>
              </Route>
            </Switch>
          </Suspense>
        </App>
      </withRouter(App)>
    </Router>
  </BrowserRouter>
</Provider>
`;
